source/Connection/AMDSClientTCPSocket.cpp

```diff
@@ -41,17 +46,38 @@ void AMDSClientTCPSocket::readFortune()
 	if (tcpSocket_->bytesAvailable() < (int)sizeof(quint32))
 		return;
 
-	quint32 blockSize;
+	AMDSDataStream *inDataStream = new AMDSDataStream(tcpSocket_);
+	inDataStream->setVersion(QDataStream::Qt_4_0);
+
+	if (!waitingMorePackages_) {
+		*inDataStream >> expectedBufferSize_;
+
+		if (tcpSocket_->bytesAvailable() < expectedBufferSize_) {
+			// more data package is expecting, we need to buffer the current ones
+			waitingMorePackages_ = true;
+			incomeDataBuffer_->clear();
+
+			readedBufferSize_ = tcpSocket_->bytesAvailable();
+			incomeDataBuffer_->append(tcpSocket_->readAll());
+		}
+	} else {
+		// more data package is coming
+		readedBufferSize_ += tcpSocket_->bytesAvailable();
+		incomeDataBuffer_->append(tcpSocket_->readAll());
+
+		if (readedBufferSize_ == expectedBufferSize_) {
+			waitingMorePackages_ = false;
 
-	AMDSDataStream inDataStream(tcpSocket_);
-	inDataStream.setVersion(QDataStream::Qt_4_0);
+			inDataStream = new AMDSDataStream(incomeDataBuffer_);
```

Comment:
@dretrex same as QByteArray, QDataStream doesn't need to force release as well. 

I refine the codes a little bit so that inDataStream won't be assigned twice when there are multiple packages